pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/DefaultRebalancePreChecker.java

```diff
@@ -0,0 +1,126 @@
+ * under the License.
+ */
+package org.apache.pinot.controller.helix.core.rebalance;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager;
+import org.apache.pinot.common.assignment.InstanceAssignmentConfigUtils;
+import org.apache.pinot.common.exception.InvalidConfigException;
+import org.apache.pinot.controller.helix.core.PinotHelixResourceManager;
+import org.apache.pinot.controller.util.TableMetadataReader;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.config.table.assignment.InstanceAssignmentConfig;
+import org.apache.pinot.spi.config.table.assignment.InstancePartitionsType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class DefaultRebalancePreChecker implements RebalancePreChecker {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultRebalancePreChecker.class);
+
+  public static final String NEEDS_RELOAD_STATUS = "needsReloadStatus";
+  public static final String IS_MINIMIZE_DATA_MOVEMENT = "isMinimizeDataMovement";
+
+  @Override
+  public Map<String, String> doRebalancePreChecks(String rebalanceJobId, String tableNameWithType,
+      TableConfig tableConfig, PinotHelixResourceManager pinotHelixResourceManager) {
+    LOGGER.info("Start pre-checks for table: {} with rebalanceJobId: {}", tableNameWithType, rebalanceJobId);
+
+    Map<String, String> preCheckResult = new HashMap<>();
+    // Check for reload status
+    Boolean needsReload = checkReloadNeededOnServers(rebalanceJobId, tableNameWithType, pinotHelixResourceManager);
+    preCheckResult.put(NEEDS_RELOAD_STATUS,
+        needsReload == null ? "error" : needsReload ? String.valueOf(true) : String.valueOf(false));
+    // Check whether minimizeDataMovement is set in TableConfig
+    boolean isMinimizeDataMovement = checkIsMinimizeDataMovement(rebalanceJobId, tableNameWithType, tableConfig);
+    preCheckResult.put(IS_MINIMIZE_DATA_MOVEMENT, String.valueOf(isMinimizeDataMovement));
+
+    LOGGER.info("End pre-checks for table: {} with rebalanceJobId: {}", tableNameWithType, rebalanceJobId);
+    return preCheckResult;
+  }
+
+  private static Boolean checkReloadNeededOnServers(String rebalanceJobId, String tableNameWithType,
+      PinotHelixResourceManager pinotHelixResourceManager) {
+    // Use at most 10 threads to get whether reload is needed from servers
+    LOGGER.info("Fetching whether reload is needed for table: {} with rebalanceJobId: {}", tableNameWithType,
+        rebalanceJobId);
+    ExecutorService executorService = Executors.newFixedThreadPool(10);
+    Boolean needsReload = null;
+    if (pinotHelixResourceManager == null) {
+      return needsReload;
```

Comment:
the executor is not closed, if this branch happens.

but better check if this `protected ExecutorService _executorService;` inside ControllerStarter can be reused, as it's like a common executor to be reused for many things. If not able to reuse it, let's name the threads for this executor, so it's easier to find them in logs to debug.

Reply:
good catch, artifact of adding code afterwards 😅 
I've modified the code to pass the `_executorService` from the `ControllerStarter` instead